# 프로그래머스 강의
github - [corazzon-github](https://github.com/corazzon/KaggleStruggle/blob/master/word2vec-nlp-tutorial/MLWave-online-learning-perceptron.ipynb)

원문 출처 : https://mlwave.com/online-learning-perceptron/

소스코드 출처

* https://github.com/MLWave/online-learning-perceptron
* https://github.com/MLWave/online-learning-perceptron/blob/master/perceptron.py
* 위 소스코드를 Python3에서 실행되도록 일부 수정하였다.
온라인 학습 퍼셉트론

* 퍼셉트론 : 가장 단순한 인공신경망이다. 이 글은 1943년에 고안된 개념에서 2015년 Kaggle 경진대회(영화 리뷰가 긍정인지 부정인지 예측)로 옮겨간다. NLP 감정 분석 작업에서 하나의 인공 신경 세포가 0.95 AUC를 얻을 수 있음을 보여준다.


## NLP with Perceptron
* McCulloch "신경 활동에 'all-or-nothing'" 활동이 있다고 추론
* 활성화 임계값에 도달하면 return 1
* 아니며 return 0
* McCulloch-Pitts 논문

example<br>
*"a Logical Calculus of the Ideas Immanent in Nervous Activity”(신경 행동에서 내재적 사고의 논리적 계산)"*<br>
새가 씨앗과 같이 작은 물체를 먹으려고 한다. 이 행동을 표로 나타내보자
![steemit](https://steemitimages.com/640x0/https://cdn.steemitimages.com/DQmPd56VKBmht266CpBn6gUUFrQhDstJPhfJaZRSEVZGRmE/noname01.png "Social Behavior Analytics")

```
1. 둘다 1이어야 1을 출력하는 것은 END 연산가 똑같다.
2. 활성화 임계값을 1.5로 설정했다고 했을 때
3. Brown과 Round가 모두 충족되었을 때만, 임계치인 1.5보다 크므로 입력값이 1이 된다. 
4. 그 외는 0으로 출력되게 된다.
```

### 퍼셉트론
퍼셉트로이 동작하는 방식은 각 노드의 가중치와 입력치를 곱한 것을 모두 합한 값이 활성화함수에 의해 판단되는데, 그 값이 임계치(보통0)보다 크면 뉴런이 활성화되고 결과값으로 1을 출력한다. 뉴런이 활성화되지 않으면 결과값으로 -1을 출력한다.

다층 퍼셉트론으로 XOR연산도 가능하다.

#### 지도학습 분류기의 일종
sum: (value1 * weight1) + (value2 * weight1) 


<h2>2017 수능 6월 평가원 모의고사 국어영역에 등장한 퍼셉트론</h2>

<pre>
<img src='https://upload.wikimedia.org/wikipedia/commons/thumb/b/bc/Neuron_Hand-tuned.svg/1179px-Neuron_Hand-tuned.svg.png' width=200>
* 이미지 출처 https://ko.wikipedia.org/wiki/%EC%8B%A0%EA%B2%BD_%EC%84%B8%ED%8F%AC

인간의 신경 조직을 수학적으로 모델링하여 컴퓨터가 인간처럼 기억․학습․판단할 수 있도록 구현한 것이 인공 신경망 기술이다. 신경 조직의 기본 단위는 뉴런인데, ⓐ 인공 신경망에서는 뉴런의 기능을 수학적으로 모델링한 퍼셉트론을 기본 단위로 사용한다.

<img src='https://upload.wikimedia.org/wikipedia/commons/thumb/3/31/Perceptron.svg/1052px-Perceptron.svg.png' width=300>
* 이미지 출처 https://en.wikipedia.org/wiki/Perceptron

ⓑ퍼셉트론은 입력값들을 받아들이는 여러 개의 ⓒ입력 단자와 이 값을 처리하는 부분, 처리된 값을 내보내는 한 개의 출력 단자로 구성되어 있다. 퍼셉트론은 각각의 입력 단자에 할당된 ⓓ 가중치를 입력값에 곱한 값들을 모두 합하여 가중합을 구한 후, 고정된 ⓔ 임계치보다 가중합이 작으면 0, 그렇지 않으면 1과 같은 방식으로 ⓕ 출력값을 내보낸다.

이러한 퍼셉트론은 출력값에 따라 두 가지로만 구분하여 입력값들을 판정할 수 있을 뿐이다. 이에 비해 복잡한 판정을 할 수 있는 인공 신경망은 다수의 퍼셉트론을 여러 계층으로 배열하여 한 계층에서 출력된 신호가 다음 계층에 있는 모든 퍼셉트론의 입력 단자에 입력값으로 입력되는 구조로 이루어진다.

이러한 인공 신경망에서 가장 처음에 입력값을 받아들이는 퍼셉트론들을 입력층, 가장 마지막에 있는 퍼셉트론들을 출력층이라고 한다.

<img src='https://upload.wikimedia.org/wikipedia/commons/thumb/f/f4/Honeycrisp.jpg/1280px-Honeycrisp.jpg' width=200>
* 이미지 출처 https://en.wikipedia.org/wiki/Apple


㉠ 어떤 사진 속 물체의 색깔과 형태로부터 그 물체가 사과인지 아닌지를 구별할 수 있도록 인공 신경망을 학습시키는 경우를 생각해 보자. 먼저 학습을 위한 입력값들 즉 학습 데이터를 만들어야 한다. 학습 데이터를 만들기 위해서는 사과 사진을 준비하고 사진에 나타난 특징인 색깔과 형태를 수치화 해야 한다. 이 경우 색깔과 형태라는 두 범주를 수치화하여 하나의 학습 데이터로 묶은 다음, ‘정답’에 해당하는 값과 함께 학습 데이터를 인공 신경망에 제공한다. 이때 같은 범주에 속하는 입력값은 동일한 입력 단자를 통해 들어가도록 해야 한다. 그리고 사과 사진에 대한 학습 데이터를 만들 때에 정답인 ‘사과이다’에 해당하는 값을 ‘1’로 설정하였다면 출력값 ‘0’은 ‘사과가 아니다’를 의미하게 된다.

인공 신경망의 작동은 크게 학습 단계와 판정 단계로 나뉜다. 학습 단계는 학습 데이터를 입력층의 입력 단자에 넣어 주고 출력층의 출력값을 구한 후, 이 출력값과 정답에 해당하는 값의 차이가 줄어들도록 가중치를 갱신하는 과정이다. 어떤 학습 데이터가 주어지면 이때의 출력값을 구하고 학습 데이터와 함께 제공된 정답에 해당하는 값에서 출력값을 뺀 값 즉 오차 값을 구한다. 이 오차 값의 일부가 출력층의 출력 단자에서 입력층의 입력 단자 방향으로 되돌아가면서 각 계층의 퍼셉트론별로 출력 신호를 만드는 데 관여한 모든 가중치들에 더해지는 방식으로 가중치들이 갱신된다. 이러한 과정을 다양한 학습 데이터에 대하여 반복하면 출력값들이 각각의 정답 값에 수렴하게 되고 판정 성능이 좋아진다. 오차 값이 0에 근접하게 되거나 가중치의 갱신이 더 이상 이루어지지 않게 되면 학습 단계를 마치고 판정 단계로 전환한다. 이때 판정의 오류를 줄이기 위해서는 학습 단계에서 대상들의 변별적 특징이 잘 반영되어 있는 서로 다른 학습 데이터를 사용하는 것이 좋다.
</pre>

### Online Learnign Perceptron in Python
* 원문 출처 : https://mlwave.com/online-learning-perceptron/

* 소스코드 출처

    * https://github.com/MLWave/online-learning-perceptron
    * https://github.com/MLWave/online-learning-perceptron/blob/master/perceptron.py
    * 위 소스코드를 Python3에서 실행되도록 일부 수정하였다.
    
    
파이썬으로 표준 라이브러리를 사용해서 위의 퍼셉트론 알고리즘을 구현했기 때문에 스크립트가 PyPy에서 실행되고 3-4배의 속도향상이 있다. 여기에 사용된 알고리즘은 Kaggle 포럼에서 처음 발견 된 tinrtgu의 온라인 로지스틱 회귀 스크립트에서 큰 영감을 얻었다고 한다.

다음 경진대회에서 Vowpal Wabbit을 통한 해시트릭을 사용하였고 코드가 공개되어 있다.

Display Advertising Challenge - Kaggle, Beat the benchmark with less then 200MB of memory.
코드 : https://kaggle2.blob.core.windows.net/forum-message-attachments/53646/1539/fast_solution.py


### 온라인 학습
online learning : https://en.wikipedia.org/wiki/Online_machine_learning
퍼셉트론은 온라인 학습이 가능하다(한 번에 하나씩 샘플을 통해 학습). 메모리에 전체 데이터 세트가 필요하지 않으므로(out-of-core) 더 큰 데이터 세트에 유용하다. 여기에서는 Vowpal Wabbit에서 온라인 학습 코드를 가져왔다.

* Vowpal Wabbit (Fast Learning)
* Hashing Representations


### 해싱 트릭
벡터화 해싱 트릭은 Vowpal Wabbit(John Langford)에서 시작되었다. 이 트릭은 퍼셉트론으로 들어오는 연결 수를 고정 된 크기로 설정한다. 고정 된 크기보다 낮은 숫자로 모든 원시 피처를 해싱한다. Vowpal Wabbit은 모든 데이터를 메모리로 읽어들이지 않고 모델을 훈련 시킬 수 있는 빠른 머신러닝 라이브러리다.


In [1]:
sample = "This movie sucks"
fixed_size = 1024

print(sample.split())

features = [(hash(f)%fixed_size,1) for f in sample.split()]

# list of tuples in form (feature_index,feature_value)
print(features)

['This', 'movie', 'sucks']
[(11, 1), (90, 1), (1010, 1)]


### 프로그레시브 검증 손실
한 번에 하나씩 표본을 학습하면 점진적으로 train loss가 된다. 모델이 타겟을 보지않고 첫 샘플을 보고 예측을 한다. 그런 다음 예측을 대상 레이블과 비교하여 오류율을 계산 한다. 오류율이 낮으면 좋은 모델에 가깝다.

### Multiple passes
Vowpal Wabbit을 사용하면 학습률이 떨어지는 데이터 집합을 여러 번 통과시킬 수 있다.

오류율이 낮아지면 데이터 세트를 여러 번 실행할 수도 있다. 트레이닝 데이터가 선형 적으로 분리 가능한 경우, 퍼셉트론은 트레이닝 세트에서 오차가 0으로 수렴된다.

In [2]:

# opts["D"] 여기에 있는 코드에서는 고정 값으로 다음의 값을 사용한다.
2 ** 25

33554432

* 위 코드에서는 pandas, numpy, scipy같은 도구를 사용하지 않고 파이썬에 내장 된 기능만을 사용해 학습하도록 한다.
* 원래 코드는 파이썬2로 되어 있지만, 파이썬3에서 돌아가도록 코드의 일부를 수정하였다.

* 소스코드 출처 https://github.com/MLWave/online-learning-perceptron

In [4]:
import re # 정규표현식을 사용해 텍스트 데이터를 정제하기 위해
import random # 랜덤 숫자를 생성하기 위해
from math import exp, log # 지수함수와 로그함수를 사용하기 위해
from datetime import datetime # 시간을 계산하기 위해
from operator import itemgetter # 키가 아닌 값으로 max, min 값을 구할 때 사용

정규표현식만을 사용해서 data clearning할 것

In [5]:
def clean(s):
    """
        Returns a cleaned, lowercased string
        
    """
    return " ".join(re.findall(r'\w+', s, flags = re.UNICODE)).lower()

In [12]:
def get_data_tsv(loc_dataset,opts):
    """
    Running through data in an online manner
    Parses a tsv file for this competition 
    and yields label, identifier and features
    output:
            label: int, The label / target (set to "1" if test set)
            id: string, the sample identifier
            features: list of tuples, 
                in the form [(hashed_feature_index,feature_value)]

    온라인 학습 방법을 통해 데이터를 실행한다.
    tsv파일을 통해 레이블, identifier, 피처(특성)를 파싱한다.
    결과물:
        label : int, 레이블 / 대상 (테스트 집합 인 경우 "1"로 설정)
        id : 문자열, 샘플 식별자
        features : [(hashed_feature_index, feature_value)] 
                형식의 튜플 목록
    """
    for e, line in enumerate(open(loc_dataset,"rb")):
        if e > 0:
            r = line.decode('utf-8').strip().split("\t")
            id = r[0]

            if opts["clean"]:
                try:
                    r[2] = clean(r[2]) # train data 일때 review
                except:
                    r[1] = clean(r[1]) # test data일 때 

            # opts["D"] = 2 ** 25 = 33554432
            # Vowpal Wabbit의 해싱트릭을 사용한다.
            # 해싱트릭은 큰 규모의 feature공간을 
            # 고정크기의 표현을 사용해 저장할 수 있게 한다.
            if len(r) == 3: # train set
                features = [(hash(f)%opts["D"],1) for f in r[2].split()]
                label = int(r[1])
            else: #test set
                features = [(hash(f)%opts["D"],1) for f in r[1].split()]
                label = 1

            # bigram을 사용하면 해당 피처[i]와 다음피처[i+1]를 함께 해싱한다.
            if opts["2grams"]:
                for i in range(len(features)-1):
                    features.append(
                        (hash(str(features[i][0])+str(features[i+1][0]))%opts["D"],1))
            yield label, id, features

In [7]:
def dot_product(features,weights):
    """
    Calculate dot product from features and weights
    input:
            features: A list of tuples [(feature_index,feature_value)]
            weights: the hashing trick weights filter, 
            note: length is max(feature_index)
    output:
            dotp: the dot product
    피처(특성)과 가중치로부터 내적을 구한다.
    """
    dotp = 0
    for f in features:
        dotp += weights[f[0]] * f[1]
    return dotp

In [8]:
def train_tron(loc_dataset,opts):
    start = datetime.now()
    print("\nPass\t\tErrors\t\tAverage\t\tNr. Samples\tSince Start")

    # 가중치 초기화
    if opts["random_init"]:
        random.seed(3003)
        weights = [random.random()] * opts["D"]
    else:
        weights = [0.] * opts["D"]

    # Running training passes
    # 학습 실행
    for pass_nr in range(opts["n_passes"]):
        error_counter = 0
        for e, (label, id, features) in enumerate( \
            get_data_tsv(loc_dataset,opts) ):

            # 퍼셉트론은 지도학습 분류기의 일종이다. 
            # 이전 값에 대한 학습으로 예측을 한다. 
            # 내적(dotproduct) 값이 임계 값보다 높거나 낮은지에 따라 
            # 초과하면 "1"을 예측하고 미만이면 "0"을 예측한다.
            dp = dot_product(features, weights) > 0.5

            # 다음 perceptron은 샘플의 레이블을 본다. 
            # 실제 레이블 데이터에서 위 퍼셉트론으로 구한 dp값을 빼준다.
            # 예측이 정확하다면, error 값은 "0"이며, 가중치만 남겨 둔다. 
            # 예측이 틀린 경우 error 값은 "1" 또는 "-1"이고 다음과 같이 가중치를 업데이트 한다.
            # weights[feature_index] += learning_rate * error * feature_value
            error = label - dp 

            # 예측이 틀린 경우 퍼셉트론은 다음과 같이 가중치를 업데이트한다.
            if error != 0:
                error_counter += 1
                # Updating the weights
                for index, value in features:
                    weights[index] += opts["learning_rate"] * error * log(1.+value)

        #Reporting stuff
        print("%s\t\t%s\t\t%s\t\t%s\t\t%s" % ( \
            pass_nr+1,
            error_counter,
            round(1 - error_counter /float(e+1),5),
            e+1,datetime.now()-start))

        #Oh heh, we have overfit :)
        if error_counter == 0 or error_counter < opts["errors_satisfied"]:
            print("%s errors found during training, halting"%error_counter)
            break
    return weights

In [9]:
def test_tron(loc_dataset,weights,opts):
    """
        output:
                preds: list, a list with
                [id,prediction,dotproduct,0-1normalized dotproduct]
    """
    start = datetime.now()
    print("\nTesting online\nErrors\t\tAverage\t\tNr. Samples\tSince Start")
    preds = [] # 예측한 값들을 넣어준다.
    error_counter = 0 # 
    for e, (label, id, features) in enumerate( \ # test데이터는 label이 없으므로 1로 초기화
        get_data_tsv(loc_dataset,opts) ):

        dotp = dot_product(features, weights)
        # 내적이 0.5보다 크다면 긍정으로 예측한다.
        dp = dotp > 0.5
        if dp > 0.5: # we predict positive class
            preds.append( [id, 1, dotp ] )
        else:
            preds.append( [id, 0, dotp ] )

        # get_data_tsv에서 테스트 데이터의 레이블을 1로 초기화 해주었음
        if label - dp != 0:
            error_counter += 1

    print("%s\t\t%s\t\t%s\t\t%s" % (
        error_counter,
        round(1 - error_counter /float(e+1),5),
        e+1,
        datetime.now()-start))

    # normalizing dotproducts between 0 and 1 
    # 내적을 구해 0과 1로 일반화 한다.
    # TODO: proper probability (bounded sigmoid?), 
    # online normalization
    max_dotp = max(preds,key=itemgetter(2))[2]
    min_dotp = min(preds,key=itemgetter(2))[2]
    for p in preds:
        # appending normalized to predictions
        # 정규화 된 값을 마지막에 추가해 준다.
        # (피처와 가중치에 대한 내적값 - 최소 내적값) / 최대 내적값 - 최소 내적값
        # 이 값이 캐글에서 0.95의 AUC를 얻을 수 있는 값이다.
        p.append((p[2]-min_dotp)/float(max_dotp-min_dotp)) 

    #Reporting stuff
    print("Done testing in %s"%str(datetime.now()-start))
    return preds

In [13]:
#Setting options
opts = {}
opts["D"] = 2 ** 25
opts["learning_rate"] = 0.1
opts["n_passes"] = 80 # Maximum number of passes to run before halting 몇번의  forloop를 돌 것인가
opts["errors_satisfied"] = 0 # Halt when training errors < errors_satisfied
opts["random_init"] = False # set random weights, else set all 0
opts["clean"] = True # clean the text a little
opts["2grams"] = True # add 2grams

#training and saving model into weights
%time weights = train_tron("../input/kaggle_data/Bag_of_Words_Meets_Bags_Of_Popcorn/labeledTrainData.tsv",opts)


Pass		Errors		Average		Nr. Samples	Since Start
1		5627		0.77492		25000		0:00:29.194759
2		3105		0.8758		25000		0:01:01.437466
3		2263		0.90948		25000		0:01:44.707877
4		1644		0.93424		25000		0:02:16.562501
5		1302		0.94792		25000		0:02:56.043091
6		1007		0.95972		25000		0:03:34.317175
7		806		0.96776		25000		0:04:11.278140
8		693		0.97228		25000		0:04:47.200586
9		522		0.97912		25000		0:05:21.177859
10		446		0.98216		25000		0:05:58.736324
11		405		0.9838		25000		0:06:39.198686
12		366		0.98536		25000		0:07:19.365113
13		228		0.99088		25000		0:07:58.637082
14		257		0.98972		25000		0:08:35.715865
15		240		0.9904		25000		0:09:11.469435
16		151		0.99396		25000		0:09:51.090158
17		155		0.9938		25000		0:10:25.319290
18		135		0.9946		25000		0:10:57.152690
19		133		0.99468		25000		0:11:29.626189
20		119		0.99524		25000		0:12:08.160479
21		130		0.9948		25000		0:12:40.789174
22		63		0.99748		25000		0:13:17.177658
23		126		0.99496		25000		0:13:53.580324
24		91		0.99636		25000		0:14:29.030079
25	

In [15]:
# testing and saving predictions into preds
%time preds = test_tron("../input/kaggle_data/Bag_of_Words_Meets_Bags_Of_Popcorn/testData.tsv",weights,opts)


Testing online
Errors		Average		Nr. Samples	Since Start
12191		0.51236		25000		0:00:22.203662
Done testing in 0:00:22.220619
Wall time: 22.2 s


In [16]:
preds[:10]

[['"12311_10"', 1, 82.13794089635356, 0.6128650904033375],
 ['"8348_2"', 0, -106.74466580623155, 0.4233657858136293],
 ['"5828_4"', 1, 9.080228065335247, 0.5395688456189145],
 ['"7186_2"', 1, 5.8224363167035165, 0.5363004172461745],
 ['"12128_7"', 1, 41.311571961372735, 0.5719054242002775],
 ['"2913_8"', 1, 83.24697638524935, 0.6139777468706531],
 ['"4396_1"', 0, -49.35207925586808, 0.48094575799721767],
 ['"395_2"', 0, 0.4158883083359156, 0.5308762169680104],
 ['"10616_1"', 0, -98.01101133117618, 0.43212795549374067],
 ['"9074_9"', 0, -27.656572504341867, 0.5027121001390813]]

In [17]:
# writing kaggle submission
# 캐글 점수 제출을 위한 서브미션 파일을 작성한다.
with open("../output/submit_perceptron.csv","wb") as outfile:
    outfile.write('"id","sentiment"\n'.encode('utf-8'))
    for p in sorted(preds):
        outfile.write("{},{}\n".format(p[0],p[3]).encode('utf-8'))

In [18]:
# 캐글에 제출할 때 바이너리 타입으로 제출하라고 되어 있는데 여기에서는 바이너리로 변환하지 않은 채 제출을 했다.
# 그런데 캐글에서 점수가 평가되고 심지어 높은 점수로! 
# 그래서 다시 바이너리 형태의 예측값을 제출했더니 점수가 낮아졌다. 
# 바이너리로 제출한 스코어가 훨씬 낮은 0.89132 가 나왔지만 튜토리얼에 있는 방법을 사용했을 때보다 훨씬 높은 스코어다.
# 머신러닝 패키지를 사용하지 않고 파이썬의 내장 기능을 통해서 이런 점수가 나올 수 있다는 것에 놀랐다.
import pandas as pd

presult = pd.DataFrame(preds)
presult.head()

0  1           2         3
0  "12311_10"  1   82.137941  0.612865
1    "8348_2"  0 -106.744666  0.423366
2    "5828_4"  1    9.080228  0.539569
3    "7186_2"  1    5.822436  0.536300
4   "12128_7"  1   41.311572  0.571905

In [19]:
output_sentiment = presult[1].value_counts()
print(output_sentiment[0] - output_sentiment[1])
output_sentiment

-618


1    12809
0    12191
Name: 1, dtype: int64

### 상위 벤치 마크
포럼에서 현재 가장 좋은 벤치 마크 는 Abhishek 의 로지스틱 회귀 스크립트다. 리더 보드에서 0.95 AUC 를 얻기 위해 학습과 테스트 세트의 메모리 안에서 tfidf-fit_transform을 사용한다. 여기에서는 이 변환을 건너 뛰고 온라인 해시 벡터 라이저를 사용한다.

Abhishek의 스크립트처럼 기능에서 2-gram(좋지 않은등)을 생성한다. 2-gram을 단순히 해쉬하고 샘플 벡터에 추가한다.

또, 문자를 소문자로 바꾸고 영숫자가 아닌 것을 제거함으로써 텍스트를 조금 더 빠르게 정제한다.

### 예측
AUC를 최적화하기 위해 1또는 0예측만 제출하지 않는다. 이것은 약 0.88의 AUC를 줄 것이다. 여기에서는 0과 1 사이에서 정규화된 dotproduct를 제출한다.

이 스크립트는 Abhishek의 솔루션 0.952 점수와 비교해 본다. 하지만 이 코드는 단지 몇 MB의 메모리를 필요로하며, tfidf 변환을 수행하지 않고 단지 2분 만에 0의 학습오류에 수렴한다다.

텍스트 정제와 2-gram이 없으면 스크립트는 60초 이내에 실행되고 0.93의 점수를 산출한다.

### 결론
절대적인 지식은 없다. 모든 것은 확률에 의해서만 이루어진다 - Gödel (1961)

0-1 임계 값 활성화가 적용된 매우 기본적인 퍼셉트론은 NLP 경진대회에서 잘 동작한다. 여기에서는 1957 알고리즘을 전혀 변경하지 않았다. 해싱 트릭을 없애고 비슷하거나 약간 더 좋은 점수를 받을 수 있다.

이 스크립트는 확률을 출력 할 수 없다. dotproduct에 bounded sigmoid를 사용하여 출력의 온라인 정규화 형태를 얻거나 tinrtgu의 (Vowpal Wabbit)스크립트를 더 보완해 볼 수 있을 것이다.

단일 노드 단일 레이어 퍼셉트론은 비선형 함수를 모델링 할 수 없으므로 더 나은 NLP 출력을 얻으려면 FFN(feedforward nets), recurrent nets, self-organizing maps, MLP(Multi-layer Perceptrons), word2vec 및 extreme learning machine (백프로파게이션이 없는 fast ffnets)을 봐야할 것이다.